src/Microsoft.OData.Client/Serialization/HttpWebRequestMessage.cs

```diff
@@ -225,12 +252,50 @@ public override string GetHeader(string headerName)
         /// <param name="headerValue">Value of the header.</param>
         public override void SetHeader(string headerName, string headerValue)
         {
-#if DEBUG
-            // Only content length header is set after firing SendingRequest2 event
-            Debug.Assert(!this.sendingRequest2Fired || headerName == XmlConstants.HttpContentLength, "!this.sendingRequest2Fired || headerName == XmlConstants.HttpContentLength");
-#endif
-            Util.CheckArgumentNullAndEmpty(headerName, "headerName");
-            HttpWebRequestMessage.SetHeaderValue(this.httpRequest, headerName, headerValue);
+            if (_contentHeaderNames.Contains(headerName))
+            {
+                // If this is a "Content" header then we cache the value (due
+                // to the message content not being set yet) and set it
+                // upon sending the message (when the content will be available)
+                if (string.Equals(headerName, XmlConstants.HttpContentType, StringComparison.OrdinalIgnoreCase))
+                {
+                    _contentHeaderValueCache[XmlConstants.HttpContentType] = headerValue;
+                }
+                else if (string.Equals(headerName, XmlConstants.HttpContentLength, StringComparison.OrdinalIgnoreCase))
+                {
+                    _contentHeaderValueCache[XmlConstants.HttpContentLength] = headerValue.ToString(CultureInfo.InvariantCulture.NumberFormat);
+                }
+                else if (string.Equals(headerName, XmlConstants.HttpContentDisposition, StringComparison.OrdinalIgnoreCase))
+                {
+                    _contentHeaderValueCache[XmlConstants.HttpContentDisposition] = headerValue.ToString(CultureInfo.InvariantCulture.NumberFormat);
+                }
+                else
+                {
+                    _contentHeaderValueCache[headerName] = headerValue;
+                }
+                return;
+            }
+
+            if (string.Equals(headerName, XmlConstants.HttpRequestAccept, StringComparison.OrdinalIgnoreCase))

```

Comment:
nit: You could move lines 279-298 to a private method `SetRequestHeader(string, string)` to keep things cleaner